### ISSUE TITLE
Persister.transaction does not rollback on panic

### ISSUE THREAD

**Issue by mitar** [ory/hydra#2857]
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/hydra/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/hydra/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Cloud](https://www.ory.sh/) project.
- [X] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

I am not sure if this is by design, but when looking at Hydra's code, [Persister.transaction](https://github.com/ory/hydra/blob/6ea0bf8f4dc990667c6911c92a1ad644733745be/persistence/sql/persister.go#L108) does not handle `panic` and does not call `c.TX.Rollback` on it. I think it would be cleaner if it would be handled here instead of potentially never handling it because panic got handled somewhere else and never killed the process/context.

### Reproducing the bug

This is a design issue.

### Version

observed on current master branch

### Additional Context

See [here an example of a similar function in Gorm](https://github.com/jinzhu/gorm/blob/master/main.go#L536-L559) which does similar handling, but handles `panic` as well.

**Comment by zepatrik** [comment]
True, it would be way cleaner to handle the rollback. Anyone working on this, please make the changes in https://github.com/ory/x/blob/master/popx/transaction.go and then use that function in Hydra instead of the current transaction implementation.

**Comment by mitar** [comment]
I am not familiar with popx, but I would suggest that this function here is changed to:

```golang
func (p *Persister) transaction(ctx context.Context, f func(ctx context.Context, c *pop.Connection) error) (err error) {
	isNested := true
        panicked := true

	c, ok := ctx.Value(transactionContextKey).(*pop.Connection)
	if !ok {
		isNested = false

		c, err = p.conn.WithContext(ctx).NewTransaction()
		if err != nil {
			return errorsx.WithStack(err)
		}
		defer func() {
			if panicked || err != nil {
				if !isNested {
					// It would be better if there was RollbackUnlessCommitted.
					rollbackErr := errorsx.WithStack(c.TX.Rollback())
					if rollbackErr != nil {
						err = rollbackErr
					}
				}
			}
		}()
	}

	err = f(context.WithValue(ctx, transactionContextKey, c), c)
	if err == nil {
		// commit if there is no wrapping transaction
		if !isNested {
			err = errorsx.WithStack(c.TX.Commit())
		}
        }

	return
}
```

**Comment by aeneasr** [comment]
Does Ory Hydra actually have a panic? If not, I'd close this as a dupe of https://github.com/ory/fosite/issues/637 and also my comment https://github.com/ory/fosite/issues/637#issuecomment-973838568

**Comment by mitar** [comment]
Are you saying that in no library Hydra is calling, they ever use panic or no code there will ever panic?

In any case, it has panic if you call into persistence layer for example, which panics. So Hydra (and fosite) are pluggable, so one cannot really fully anticipate behavior of that code. Moreover, Go panics also when there are other issues, like nil pointer dereferencing or interface assertion mismatch.

**Comment by mitar** [comment]
> Are you saying that in no library Hydra is calling, they ever use panic or no code there will ever panic?

Hydra calls into pop, which calls into sglx, which has many uses of `panic`:

https://github.com/jmoiron/sqlx/search?q=panic

**Comment by aeneasr** [comment]
What I meant was if you actually have observed a panic or if this is a hypothetical fix. Please also do keep in mind that if we begin a transaction, the transaction is not "commited" until you actually do "commit". So a panic will automagically rollback the transaction if the connection is closed (on panic). So I don't really understand what bug case this would solve? Or did I miss a bug here (which is why I asked)?

**Comment by mitar** [comment]
I have observed panics while developing our own storage backend and when I had bugs. Hopefully I have ironed them out, but I also observed that it didn't cleanup things well. So I worry that if I have any more bug around, things might be problematic in production.

> So a panic will automagically rollback the transaction if the connection is closed (on panic).

Unless you are using connection pooling, then transaction might be kept around and the next query hitting the same connection might be in for a surprise.

**Comment by aeneasr** [comment]
> Unless you are using connection pooling, then transaction might be kept around and the next query hitting the same connection might be in for a surprise.

Oh now I see the issue, @zepatrik also made me aware that the problem is connection pooling in combination with panic recovery in the middleware, in combination with a panic in the storage adapter. The good news is, in Ory Hydra we do not have panic recovery because we actually want the system to crash in order to have a high impact issue which must be resolved, as panics are always very bad due to lack of controlled execution flow! Still possible to address this in Fosite though. As I said already though, the refactor will be substantial and it only covers very rare edge cases, so it's a risk / reward decision that has to be made.

**Comment by mitar** [comment]
> The good news is, in Ory Hydra we do not have panic recovery because we actually want the system to crash in order to have a high impact issue which must be resolved, as panics are always very bad due to lack of controlled execution flow!

Oh, I am not saying that you should handle the panic, only that you should call rollback in defer and then keep the panic propagating. I understand that killing the process probably addresses the connection pooling issue (although not sure what happens if somebody is using pg_bouncer, I hope it cleans up the transaction automatically when connection dies), but I think that it could help with storage backends which are not transactional (like mongo) but one wants to take advantage of transaction API to manually cleanup things. But that is pure speculation.

At least for Hydra, the change necessary to call rollback on panic is relatively small, I posted it above. Not sure why we would not do that. For Fosite, I can look into that a bit and see how much work it really is.

**Comment by aeneasr** [comment]
Ok, sounds good .)

**Comment by mitar** [comment]
Done for Fosite in https://github.com/ory/fosite/pull/638.

**Comment by mitar** [comment]
So I made this for Fosite, but for Hydra the change still has to be done, see: https://github.com/ory/hydra/issues/2857#issuecomment-969435112

**Comment by kmherrmann** [comment]
thanks, reopening!

**Comment by alnr** [comment]
Thanks for the patience on this one -- and for pushing the equivalent change into fosite back in 2021. Both halves of what @zepatrik asked for are done now:

- `ory/x/popx` does the rollback-then-repanic you proposed, in [`popx/transaction.go`](https://github.com/ory/x/blob/master/popx/transaction.go) (postgres/mysql/yugabyte path):

```go
defer func() {
    if ex := recover(); ex != nil {
        _ = cn.TX.Rollback()
        panic(ex)
    }
}()
```

- And Hydra uses it: `persistence/sql/persister.go` is now just `return popx.Transaction(ctx, p.c, f)`.

The SQLite/default path goes through `pop.Connection.Transaction`, which has its own `recover()` + rollback, and CockroachDB goes through `crdb.ExecuteInTx`. So every dialect rolls back on panic and the panic still propagates -- exactly the behaviour you asked for, including your point about not swallowing it.

Closing as fixed. Thanks again for sticking with this!
